{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haiyi/anaconda3/envs/halo2/lib/python3.7/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from typing import Dict, Iterable, Optional\n",
    "\n",
    "import numpy as np\n",
    "import torch\n",
    "from torch.distributions import Normal, Poisson\n",
    "from torch.distributions import kl_divergence as kld\n",
    "from torch import tensor\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import tools\n",
    "import logging\n",
    "# torch.autograd.set_det"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haiyi/anaconda3/envs/halo2/lib/python3.7/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_3094399_3095523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_3113503_3114077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_3119414_3121804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_3181343_3181401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_3198296_3198611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   peak gene distance   peak_type\n",
       "0  chr1_3094399_3095523  NaN      NaN  intergenic\n",
       "1  chr1_3113503_3114077  NaN      NaN  intergenic\n",
       "2  chr1_3119414_3121804  NaN      NaN  intergenic\n",
       "3  chr1_3181343_3181401  NaN      NaN  intergenic\n",
       "4  chr1_3198296_3198611  NaN      NaN  intergenic"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## load RNA data and ATAC data and ATAC annotation data\n",
    "rna_data = anndata.read_h5ad(\"data/datasets/mouse_brain_rna.h5ad\")\n",
    "atac_data = anndata.read_h5ad(\"data/datasets/mouse_brain_atac.h5ad\")\n",
    "atac_annotation = pd.read_csv(\"data/datasets/e18_mouse_brain_fresh_5k_atac_peak_annotation.tsv\", sep='\\t')\n",
    "atac_annotation.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Xkr4               True\n",
       "Gm1992            False\n",
       "Gm19938           False\n",
       "Rp1                True\n",
       "Mrpl15            False\n",
       "                  ...  \n",
       "CAAA01118383.1    False\n",
       "Vamp7             False\n",
       "Tmlhe              True\n",
       "4933409K07Rik     False\n",
       "AC149090.1        False\n",
       "Name: exog, Length: 14583, dtype: bool"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## preprocessing RNA data\n",
    "sc.pp.filter_genes(rna_data, min_cells=15)\n",
    "rna_data.raw = rna_data\n",
    "sc.pp.normalize_total(rna_data, target_sum=1e4)\n",
    "sc.pp.log1p(rna_data)\n",
    "sc.pp.highly_variable_genes(rna_data, min_disp = 0.2)\n",
    "rna_data.var['exog'] = rna_data.var.highly_variable.copy()\n",
    "rna_data.var['endog'] = rna_data.var.exog & (rna_data.var.dispersions_norm > 0.7)\n",
    "rna_data.layers['counts'] = rna_data.raw.to_adata().X.copy()\n",
    "rna_data.var['exog']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_peakid(row):\n",
    "    chr = row['chr']\n",
    "    start = row[\"start\"]\n",
    "    end = row[\"end\"]\n",
    "    peak = chr+'_' + start+ '_' + end\n",
    "    return peak\n",
    "\n",
    "\n",
    "def get_atac_exog(row, rna_dict):\n",
    "\n",
    "    exog = False\n",
    "    # endo = False\n",
    "    genes = str(row[\"gene\"])\n",
    "    if genes == 'nan':\n",
    "        return exog\n",
    "    else:\n",
    "        genes = genes.split(\";\")\n",
    "        for g in genes:\n",
    "            if g in rna_dict['exog'].keys():\n",
    "                if rna_dict['exog'][g]:\n",
    "                    exog = True\n",
    "            # if g in rna_dict['endog'].keys():        \n",
    "            #     if  rna_dict['endog'][g]:\n",
    "            #         endo = True\n",
    "\n",
    "    return exog\n",
    "\n",
    "\n",
    "def get_atac_endog(row, rna_dict):\n",
    "\n",
    "    endo = False\n",
    "    genes = str(row[\"gene\"])\n",
    "    if genes == 'nan':\n",
    "        return endo\n",
    "    else:\n",
    "        genes = genes.split(\";\")\n",
    "        for g in genes:\n",
    "    \n",
    "            if g in rna_dict['endog'].keys():        \n",
    "                if  rna_dict['endog'][g]:\n",
    "                    endo = True\n",
    "                \n",
    "\n",
    "    return endo    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3094399</td>\n",
       "      <td>3095523</td>\n",
       "      <td>235</td>\n",
       "      <td>chr1_3094399_3095523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3113503</td>\n",
       "      <td>3114077</td>\n",
       "      <td>103</td>\n",
       "      <td>chr1_3113503_3114077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3119414</td>\n",
       "      <td>3121804</td>\n",
       "      <td>449</td>\n",
       "      <td>chr1_3119414_3121804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3198296</td>\n",
       "      <td>3198611</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1_3198296_3198611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3210204</td>\n",
       "      <td>3210605</td>\n",
       "      <td>52</td>\n",
       "      <td>chr1_3210204_3210605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  peak_id   chr    start      end  n_cells                  peak\n",
       "0       1  chr1  3094399  3095523      235  chr1_3094399_3095523\n",
       "1       2  chr1  3113503  3114077      103  chr1_3113503_3114077\n",
       "2       3  chr1  3119414  3121804      449  chr1_3119414_3121804\n",
       "3       4  chr1  3198296  3198611       46  chr1_3198296_3198611\n",
       "4       5  chr1  3210204  3210605       52  chr1_3210204_3210605"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_data.var['peak'] =  atac_data.var.apply(generate_peakid, axis=1)\n",
    "atac_data.var = atac_data.var.reset_index()\n",
    "atac_data.var.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haiyi/anaconda3/envs/halo2/lib/python3.7/site-packages/anndata/_core/anndata.py:798: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3094399</td>\n",
       "      <td>3095523</td>\n",
       "      <td>235</td>\n",
       "      <td>chr1_3094399_3095523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3113503</td>\n",
       "      <td>3114077</td>\n",
       "      <td>103</td>\n",
       "      <td>chr1_3113503_3114077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3119414</td>\n",
       "      <td>3121804</td>\n",
       "      <td>449</td>\n",
       "      <td>chr1_3119414_3121804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3198296</td>\n",
       "      <td>3198611</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1_3198296_3198611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3210204</td>\n",
       "      <td>3210605</td>\n",
       "      <td>52</td>\n",
       "      <td>chr1_3210204_3210605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3292586</td>\n",
       "      <td>3292976</td>\n",
       "      <td>45</td>\n",
       "      <td>chr1_3292586_3292976</td>\n",
       "      <td>Xkr4;Gm1992</td>\n",
       "      <td>0;-173611</td>\n",
       "      <td>distal;distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3371598</td>\n",
       "      <td>3371961</td>\n",
       "      <td>53</td>\n",
       "      <td>chr1_3371598_3371961</td>\n",
       "      <td>Xkr4;Gm1992</td>\n",
       "      <td>0;-94626</td>\n",
       "      <td>distal;distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3399683</td>\n",
       "      <td>3400422</td>\n",
       "      <td>132</td>\n",
       "      <td>chr1_3399683_3400422</td>\n",
       "      <td>Xkr4;Gm1992</td>\n",
       "      <td>0;-66165</td>\n",
       "      <td>distal;distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3477039</td>\n",
       "      <td>3477771</td>\n",
       "      <td>87</td>\n",
       "      <td>chr1_3477039_3477771</td>\n",
       "      <td>Xkr4;Gm1992</td>\n",
       "      <td>0;10453</td>\n",
       "      <td>distal;distal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3508598</td>\n",
       "      <td>3509041</td>\n",
       "      <td>55</td>\n",
       "      <td>chr1_3508598_3509041</td>\n",
       "      <td>Xkr4;Gm1992</td>\n",
       "      <td>0;42012</td>\n",
       "      <td>distal;distal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  peak_id   chr    start      end  n_cells                  peak         gene  \\\n",
       "0       1  chr1  3094399  3095523      235  chr1_3094399_3095523          NaN   \n",
       "1       2  chr1  3113503  3114077      103  chr1_3113503_3114077          NaN   \n",
       "2       3  chr1  3119414  3121804      449  chr1_3119414_3121804          NaN   \n",
       "3       4  chr1  3198296  3198611       46  chr1_3198296_3198611          NaN   \n",
       "4       5  chr1  3210204  3210605       52  chr1_3210204_3210605          NaN   \n",
       "5       6  chr1  3292586  3292976       45  chr1_3292586_3292976  Xkr4;Gm1992   \n",
       "6       7  chr1  3371598  3371961       53  chr1_3371598_3371961  Xkr4;Gm1992   \n",
       "7       8  chr1  3399683  3400422      132  chr1_3399683_3400422  Xkr4;Gm1992   \n",
       "8       9  chr1  3477039  3477771       87  chr1_3477039_3477771  Xkr4;Gm1992   \n",
       "9      10  chr1  3508598  3509041       55  chr1_3508598_3509041  Xkr4;Gm1992   \n",
       "\n",
       "    distance      peak_type  \n",
       "0        NaN     intergenic  \n",
       "1        NaN     intergenic  \n",
       "2        NaN     intergenic  \n",
       "3        NaN     intergenic  \n",
       "4        NaN     intergenic  \n",
       "5  0;-173611  distal;distal  \n",
       "6   0;-94626  distal;distal  \n",
       "7   0;-66165  distal;distal  \n",
       "8    0;10453  distal;distal  \n",
       "9    0;42012  distal;distal  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_var = pd.merge(atac_data.var, atac_annotation, left_on='peak', right_on='peak')\n",
    "atac_data.var = new_var\n",
    "atac_data.var.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_dict = rna_data.var.to_dict()\n",
    "newvar = atac_data.var\n",
    "newvar['exog'] =  newvar.apply(get_atac_exog, rna_dict=rna_dict,  axis=1)\n",
    "newvar['endog'] =  newvar.apply(get_atac_endog, rna_dict=rna_dict,  axis=1)\n",
    "atac_data.var = newvar.set_index('peak_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>peak</th>\n",
       "      <th>gene</th>\n",
       "      <th>distance</th>\n",
       "      <th>peak_type</th>\n",
       "      <th>exog</th>\n",
       "      <th>endog</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peak_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3094399</td>\n",
       "      <td>3095523</td>\n",
       "      <td>235</td>\n",
       "      <td>chr1_3094399_3095523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3113503</td>\n",
       "      <td>3114077</td>\n",
       "      <td>103</td>\n",
       "      <td>chr1_3113503_3114077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3119414</td>\n",
       "      <td>3121804</td>\n",
       "      <td>449</td>\n",
       "      <td>chr1_3119414_3121804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3198296</td>\n",
       "      <td>3198611</td>\n",
       "      <td>46</td>\n",
       "      <td>chr1_3198296_3198611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3210204</td>\n",
       "      <td>3210605</td>\n",
       "      <td>52</td>\n",
       "      <td>chr1_3210204_3210605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123879</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>37994</td>\n",
       "      <td>39286</td>\n",
       "      <td>266</td>\n",
       "      <td>GL456216.1_37994_39286</td>\n",
       "      <td>CAAA01118383.1</td>\n",
       "      <td>22114</td>\n",
       "      <td>distal</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123880</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>40534</td>\n",
       "      <td>41810</td>\n",
       "      <td>139</td>\n",
       "      <td>GL456216.1_40534_41810</td>\n",
       "      <td>CAAA01118383.1</td>\n",
       "      <td>24654</td>\n",
       "      <td>distal</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123881</th>\n",
       "      <td>GL456216.1</td>\n",
       "      <td>49045</td>\n",
       "      <td>49376</td>\n",
       "      <td>35</td>\n",
       "      <td>GL456216.1_49045_49376</td>\n",
       "      <td>CAAA01118383.1</td>\n",
       "      <td>33165</td>\n",
       "      <td>distal</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123882</th>\n",
       "      <td>JH584292.1</td>\n",
       "      <td>12573</td>\n",
       "      <td>14868</td>\n",
       "      <td>231</td>\n",
       "      <td>JH584292.1_12573_14868</td>\n",
       "      <td>Vmn2r122</td>\n",
       "      <td>9001</td>\n",
       "      <td>distal</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123883</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1976</td>\n",
       "      <td>835</td>\n",
       "      <td>JH584295.1_0_1976</td>\n",
       "      <td>CAAA01147332.1</td>\n",
       "      <td>0</td>\n",
       "      <td>promoter</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123883 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr    start      end  n_cells                    peak  \\\n",
       "peak_id                                                                  \n",
       "1              chr1  3094399  3095523      235    chr1_3094399_3095523   \n",
       "2              chr1  3113503  3114077      103    chr1_3113503_3114077   \n",
       "3              chr1  3119414  3121804      449    chr1_3119414_3121804   \n",
       "4              chr1  3198296  3198611       46    chr1_3198296_3198611   \n",
       "5              chr1  3210204  3210605       52    chr1_3210204_3210605   \n",
       "...             ...      ...      ...      ...                     ...   \n",
       "123879   GL456216.1    37994    39286      266  GL456216.1_37994_39286   \n",
       "123880   GL456216.1    40534    41810      139  GL456216.1_40534_41810   \n",
       "123881   GL456216.1    49045    49376       35  GL456216.1_49045_49376   \n",
       "123882   JH584292.1    12573    14868      231  JH584292.1_12573_14868   \n",
       "123883   JH584295.1        0     1976      835       JH584295.1_0_1976   \n",
       "\n",
       "                   gene distance   peak_type   exog  endog  \n",
       "peak_id                                                     \n",
       "1                   NaN      NaN  intergenic  False  False  \n",
       "2                   NaN      NaN  intergenic  False  False  \n",
       "3                   NaN      NaN  intergenic  False  False  \n",
       "4                   NaN      NaN  intergenic  False  False  \n",
       "5                   NaN      NaN  intergenic  False  False  \n",
       "...                 ...      ...         ...    ...    ...  \n",
       "123879   CAAA01118383.1    22114      distal  False  False  \n",
       "123880   CAAA01118383.1    24654      distal  False  False  \n",
       "123881   CAAA01118383.1    33165      distal  False  False  \n",
       "123882         Vmn2r122     9001      distal  False  False  \n",
       "123883   CAAA01147332.1        0    promoter  False  False  \n",
       "\n",
       "[123883 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_data.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "123883it [00:01, 62607.47it/s]\n"
     ]
    }
   ],
   "source": [
    "## get the motif\n",
    "tools.motif_scan.logger.setLevel(logging.INFO)\n",
    "tools.get_motif_hits_in_peaks(atac_data,\n",
    "                    genome_fasta='mm10.fa',\n",
    "                    chrom = 'chr', start = 'start', end = 'end')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_data =  anndata.read_h5ad(\"data/datasets/mouse_brain_atac_motifs2.h5ad\")\n",
    "rna_data =  anndata.read_h5ad(\"data/datasets/mouse_brain_rna_exn.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>parsed_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0973.1</td>\n",
       "      <td>CDF2</td>\n",
       "      <td>CDF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA1274.1</td>\n",
       "      <td>OBP3</td>\n",
       "      <td>OBP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA1559.1</td>\n",
       "      <td>SNAI3</td>\n",
       "      <td>SNAI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0068.2</td>\n",
       "      <td>PAX4</td>\n",
       "      <td>PAX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA1063.1</td>\n",
       "      <td>TCP19</td>\n",
       "      <td>TCP19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641</th>\n",
       "      <td>MA0999.1</td>\n",
       "      <td>ERF098</td>\n",
       "      <td>ERF098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1642</th>\n",
       "      <td>MA0930.1</td>\n",
       "      <td>ABF3</td>\n",
       "      <td>ABF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1643</th>\n",
       "      <td>MA1589.1</td>\n",
       "      <td>ZNF140</td>\n",
       "      <td>ZNF140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1644</th>\n",
       "      <td>MA1007.1</td>\n",
       "      <td>PHYPADRAFT</td>\n",
       "      <td>PHYPADRAFT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645</th>\n",
       "      <td>MA0030.1</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>FOXF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1646 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id        name parsed_name\n",
       "0     MA0973.1        CDF2        CDF2\n",
       "1     MA1274.1        OBP3        OBP3\n",
       "2     MA1559.1       SNAI3       SNAI3\n",
       "3     MA0068.2        PAX4        PAX4\n",
       "4     MA1063.1       TCP19       TCP19\n",
       "...        ...         ...         ...\n",
       "1641  MA0999.1      ERF098      ERF098\n",
       "1642  MA0930.1        ABF3        ABF3\n",
       "1643  MA1589.1      ZNF140      ZNF140\n",
       "1644  MA1007.1  PHYPADRAFT  PHYPADRAFT\n",
       "1645  MA0030.1       FOXF2       FOXF2\n",
       "\n",
       "[1646 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tools.fetch_factor_meta(atac_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/haiyi/Desktop/HierachicalCausal_Omics/tools/adata_interface/utils.py:226: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  X = hits_matrix,\n",
      "/home/haiyi/anaconda3/envs/halo2/lib/python3.7/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1646 × 123883\n",
       "    obs: 'id', 'name', 'parsed_name'\n",
       "    var: 'chr', 'start', 'end', 'n_cells', 'peak', 'gene', 'distance', 'peak_type', 'exog', 'endog'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tools.fetch_factor_hits(atac_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3365 × 123883\n",
       "    obs: 'celltype', 'batch_id', 'modality', 'latent_time', '_scvi_batch', '_scvi_labels', 'latent_decouple', 'latent_couple', 'latent_1', 'latent_2', 'latent_3', 'latent_4', 'latent_5', 'latent_6', 'latent_7', 'latent_8', 'latent_9', 'latent_10'\n",
       "    var: 'chr', 'start', 'end', 'n_cells', 'peak', 'gene', 'distance', 'peak_type', 'exog', 'endog'\n",
       "    uns: 'motifs', 'neighbors', 'umap'\n",
       "    obsm: 'X_umap', 'latent'\n",
       "    varm: 'motifs_hits'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_data.write_h5ad(\"data/datasets/mouse_brain_atac_motifs2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>in_expr_data</th>\n",
       "      <th>name</th>\n",
       "      <th>parsed_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0973.1</td>\n",
       "      <td>True</td>\n",
       "      <td>CDF2</td>\n",
       "      <td>CDF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA1274.1</td>\n",
       "      <td>True</td>\n",
       "      <td>OBP3</td>\n",
       "      <td>OBP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA1559.1</td>\n",
       "      <td>True</td>\n",
       "      <td>SNAI3</td>\n",
       "      <td>SNAI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0068.2</td>\n",
       "      <td>True</td>\n",
       "      <td>PAX4</td>\n",
       "      <td>PAX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA1063.1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCP19</td>\n",
       "      <td>TCP19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641</th>\n",
       "      <td>MA0999.1</td>\n",
       "      <td>True</td>\n",
       "      <td>ERF098</td>\n",
       "      <td>ERF098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1642</th>\n",
       "      <td>MA0930.1</td>\n",
       "      <td>True</td>\n",
       "      <td>ABF3</td>\n",
       "      <td>ABF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1643</th>\n",
       "      <td>MA1589.1</td>\n",
       "      <td>True</td>\n",
       "      <td>ZNF140</td>\n",
       "      <td>ZNF140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1644</th>\n",
       "      <td>MA1007.1</td>\n",
       "      <td>True</td>\n",
       "      <td>PHYPADRAFT</td>\n",
       "      <td>PHYPADRAFT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645</th>\n",
       "      <td>MA0030.1</td>\n",
       "      <td>True</td>\n",
       "      <td>FOXF2</td>\n",
       "      <td>FOXF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1646 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id  in_expr_data        name parsed_name\n",
       "0     MA0973.1          True        CDF2        CDF2\n",
       "1     MA1274.1          True        OBP3        OBP3\n",
       "2     MA1559.1          True       SNAI3       SNAI3\n",
       "3     MA0068.2          True        PAX4        PAX4\n",
       "4     MA1063.1          True       TCP19       TCP19\n",
       "...        ...           ...         ...         ...\n",
       "1641  MA0999.1          True      ERF098      ERF098\n",
       "1642  MA0930.1          True        ABF3        ABF3\n",
       "1643  MA1589.1          True      ZNF140      ZNF140\n",
       "1644  MA1007.1          True  PHYPADRAFT  PHYPADRAFT\n",
       "1645  MA0030.1          True       FOXF2       FOXF2\n",
       "\n",
       "[1646 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(atac_data.uns['motifs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools.subset_factors(atac_data,\n",
    "                          use_factors=[factor for factor in rna_data.var_names\n",
    "                                       if not ('FOS' in factor or 'JUN' in factor)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_data.write_h5ad(\"data/datasets/mouse_brain_atac_motifs2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<123883x1646 sparse matrix of type '<class 'numpy.float64'>'\n",
       "\twith 39535877 stored elements in Compressed Sparse Column format>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_data.varm['motifs_hits']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3365 × 123883\n",
       "    obs: 'celltype', 'batch_id', 'modality', 'latent_time', '_scvi_batch', '_scvi_labels', 'latent_decouple', 'latent_couple', 'latent_1', 'latent_2', 'latent_3', 'latent_4', 'latent_5', 'latent_6', 'latent_7', 'latent_8', 'latent_9', 'latent_10'\n",
       "    var: 'chr', 'start', 'end', 'n_cells', 'peak', 'gene', 'distance', 'peak_type', 'exog', 'endog'\n",
       "    uns: 'motifs', 'neighbors', 'umap'\n",
       "    obsm: 'X_umap', 'latent'\n",
       "    varm: 'motifs_hits'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### MIRA Test\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.16 ('halo2')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b1a2dff3d0af1d2ba6092416114f7f000e7e116c16a32271780d730fd43a1ad4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
