{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import sys\n",
    "import multivelo as mv\n",
    "import scanpy as sc\n",
    "import scvelo as scv\n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "# from dtw import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.settings.verbosity = 3\n",
    "scv.settings.presenter_view = True\n",
    "scv.set_figure_params('scvelo')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 200)\n",
    "np.set_printoptions(suppress=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "rna_path = \"/mnt/data0/halo/mouse_breastcancer/C1\"\n",
    "# atac_path = \"data/datasets/human_brain_atac_gene_after_filt_r2.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_c1 = sc.read_10x_mtx(rna_path, gex_only=False)\n",
    "# adata_atac = sc.read(atac_path, =atac_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16207 × 173854\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_c1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACGCGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAAGCTCC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACACCTA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAATCCTGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAATTGACT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGCTGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAGTACG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTCAGGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGCTGT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGCGG-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16207 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACAGCCAAACGCGA-1, AAACAGCCAAAGCTCC-1, AAACAGCCAACACCTA-1, AAACAGCCAATCCTGA-1, AAACAGCCAATTGACT-1, AAACAGCCAATTTGGT-1, AAACAGCCACACTAAT-1, AAACAGCCAGCATGAG-1, AAACAGCCAGGCTGTT-1, AAACAGCCAGTTATGT-1, AAACAGCCATACCCGG-1, AAACAGCCATAGCTGC-1, AAACAGCCATATTGAC-1, AAACAGCCATCGTTCT-1, AAACAGCCATGAATAG-1, AAACAGCCATGTGGGA-1, AAACAGCCATTAAGTC-1, AAACATGCACCTAATG-1, AAACATGCACGAATCC-1, AAACATGCAGGCTACT-1, AAACATGCATCATGGC-1, AAACATGCATCCTAGA-1, AAACATGCATTTAAGC-1, AAACCAACAAACCTAT-1, AAACCAACAAACTGTT-1, AAACCAACAACAGGTG-1, AAACCAACAAGGGTTG-1, AAACCAACAATTATGC-1, AAACCAACACATACTG-1, AAACCAACACTAAGTT-1, AAACCAACAGCATGAG-1, AAACCAACAGCATGGA-1, AAACCAACATCATGTG-1, AAACCAACATCTAGCA-1, AAACCGAAGAGAAGGG-1, AAACCGAAGCTACTGG-1, AAACCGAAGCTATTGA-1, AAACCGAAGGAACACA-1, AAACCGAAGGACACTT-1, AAACCGAAGGTCGATT-1, AAACCGAAGGTGCGGA-1, AAACCGAAGGTGTTAC-1, AAACCGAAGTAAAGGT-1, AAACCGAAGTTGGGCC-1, AAACCGCGTCAGGCAT-1, AAACCGCGTTAACGAT-1, AAACCGCGTTAGTTGG-1, AAACCGCGTTCGCTTG-1, AAACCGGCAACAGGAT-1, AAACCGGCAAGGGTTG-1, AAACCGGCAATCCTGA-1, AAACCGGCAGGTCCTG-1, AAACCGGCAGGTTACC-1, AAACCGGCATGTTTGG-1, AAACGCGCAAACCTTG-1, AAACGCGCAACCTGGT-1, AAACGCGCAAGTGTTT-1, AAACGCGCAAGTTATC-1, AAACGCGCAATAACGA-1, AAACGCGCAATAAGCA-1, AAACGCGCACCCTCAC-1, AAACGCGCACCTCGCT-1, AAACGCGCAGTTTCTC-1, AAACGCGCATCTTGAG-1, AAACGGATCACGAATC-1, AAACGGATCATGGCTG-1, AAACGGATCATGTGGT-1, AAACGGATCCCTGACT-1, AAACGGATCGAAGTAG-1, AAACGGATCGTGCCTT-1, AAACGGATCGTTATAG-1, AAACGGATCTACTAAG-1, AAACGGATCTTAATGG-1, AAACGGATCTTTGAGA-1, AAACGTACAACGTGCT-1, AAACGTACAACTAACT-1, AAACGTACAAGCCAGA-1, AAACGTACACTAGCGT-1, AAACGTACACTTAGGC-1, AAACGTACAGCAAGAT-1, AAACGTACAGGAACCA-1, AAACGTACAGGTCCTG-1, AAACGTACAGTTTCTC-1, AAACGTACATAGCTGC-1, AAACGTACATCTTGAG-1, AAACGTACATTGCGGT-1, AAAGCAAGTACCTTAC-1, AAAGCAAGTATACTGG-1, AAAGCACCAAAGCGGC-1, AAAGCACCAACTAGGG-1, AAAGCACCAAGGTCCT-1, AAAGCACCACCAGGTT-1, AAAGCACCACTAAATC-1, AAAGCACCACTGACTA-1, AAAGCACCACTTAACG-1, AAAGCACCAGCTCAAC-1, AAAGCACCAGGTTACC-1, AAAGCACCAGTTTGTG-1, AAAGCACCATAATCGT-1, AAAGCACCATAATTGC-1, ...]\n",
       "\n",
       "[16207 rows x 0 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_c1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3793 × 101666\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_path = \"/mnt/data0/halo/mouse_breastcancer/C2\"\n",
    "adata_c2 = sc.read_10x_mtx(rna_path, gex_only=False)\n",
    "adata_c2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>ENSMUSG00000089699</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm19938</th>\n",
       "      <td>ENSMUSG00000102331</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37381</th>\n",
       "      <td>ENSMUSG00000102343</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL456216.1:25495-26401</th>\n",
       "      <td>GL456216.1:25495-26401</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL456216.1:40355-41127</th>\n",
       "      <td>GL456216.1:40355-41127</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL456216.1:48833-49726</th>\n",
       "      <td>GL456216.1:48833-49726</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL456216.1:50856-51892</th>\n",
       "      <td>GL456216.1:50856-51892</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JH584292.1:13990-14917</th>\n",
       "      <td>JH584292.1:13990-14917</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101666 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      gene_ids    feature_types\n",
       "Xkr4                        ENSMUSG00000051951  Gene Expression\n",
       "Gm1992                      ENSMUSG00000089699  Gene Expression\n",
       "Gm19938                     ENSMUSG00000102331  Gene Expression\n",
       "Gm37381                     ENSMUSG00000102343  Gene Expression\n",
       "Rp1                         ENSMUSG00000025900  Gene Expression\n",
       "...                                        ...              ...\n",
       "GL456216.1:25495-26401  GL456216.1:25495-26401            Peaks\n",
       "GL456216.1:40355-41127  GL456216.1:40355-41127            Peaks\n",
       "GL456216.1:48833-49726  GL456216.1:48833-49726            Peaks\n",
       "GL456216.1:50856-51892  GL456216.1:50856-51892            Peaks\n",
       "JH584292.1:13990-14917  JH584292.1:13990-14917            Peaks\n",
       "\n",
       "[101666 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_c2.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_path = \"/mnt/data0/halo/mouse_breastcancer/GSM6533606_filtered_peak_bc_matrix_barcodes.tsv.gz\"\n",
    "barcode = pd.read_csv(barcode_path, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACGAAAGAACAGGA-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAACATAGCCAT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACGAAGTAACTCCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAGTATGCTAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTGACGCAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTTATAGAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4512</th>\n",
       "      <td>TTTGTGTGTGTCGTGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4513</th>\n",
       "      <td>TTTGTGTGTTCTACCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4514</th>\n",
       "      <td>TTTGTGTTCATTCATC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4515</th>\n",
       "      <td>TTTGTGTTCCCTACGT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4516</th>\n",
       "      <td>TTTGTGTTCTACTTTG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4517 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AAACGAAAGAACAGGA-1\n",
       "0     AAACGAACATAGCCAT-1\n",
       "1     AAACGAAGTAACTCCA-1\n",
       "2     AAACGAAGTATGCTAC-1\n",
       "3     AAACGAAGTGACGCAA-1\n",
       "4     AAACGAAGTTATAGAG-1\n",
       "...                  ...\n",
       "4512  TTTGTGTGTGTCGTGC-1\n",
       "4513  TTTGTGTGTTCTACCC-1\n",
       "4514  TTTGTGTTCATTCATC-1\n",
       "4515  TTTGTGTTCCCTACGT-1\n",
       "4516  TTTGTGTTCTACTTTG-1\n",
       "\n",
       "[4517 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>gex_barcode</th>\n",
       "      <th>atac_barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>excluded_reason</th>\n",
       "      <th>gex_raw_reads</th>\n",
       "      <th>gex_mapped_reads</th>\n",
       "      <th>gex_conf_intergenic_reads</th>\n",
       "      <th>gex_conf_exonic_reads</th>\n",
       "      <th>gex_conf_intronic_reads</th>\n",
       "      <th>...</th>\n",
       "      <th>atac_raw_reads</th>\n",
       "      <th>atac_unmapped_reads</th>\n",
       "      <th>atac_lowmapq</th>\n",
       "      <th>atac_dup_reads</th>\n",
       "      <th>atac_chimeric_reads</th>\n",
       "      <th>atac_mitochondrial_reads</th>\n",
       "      <th>atac_fragments</th>\n",
       "      <th>atac_TSS_fragments</th>\n",
       "      <th>atac_peak_region_fragments</th>\n",
       "      <th>atac_peak_region_cutsites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAGCCAAACAACA-1</td>\n",
       "      <td>AAACAGCCAAACAACA-1</td>\n",
       "      <td>ACAGCGGGTGTGTTAC-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACAGCCAAACATAG-1</td>\n",
       "      <td>AAACAGCCAAACATAG-1</td>\n",
       "      <td>ACAGCGGGTTGTTCTT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>33621</td>\n",
       "      <td>32008</td>\n",
       "      <td>3677</td>\n",
       "      <td>6527</td>\n",
       "      <td>19433</td>\n",
       "      <td>...</td>\n",
       "      <td>16733</td>\n",
       "      <td>212</td>\n",
       "      <td>2053</td>\n",
       "      <td>4694</td>\n",
       "      <td>28</td>\n",
       "      <td>1836</td>\n",
       "      <td>7901</td>\n",
       "      <td>3107</td>\n",
       "      <td>3586</td>\n",
       "      <td>6997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCCAAACCCTA-1</td>\n",
       "      <td>AAACAGCCAAACCCTA-1</td>\n",
       "      <td>ACAGCGGGTAACAGGC-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACAGCCAAACCTAT-1</td>\n",
       "      <td>AAACAGCCAAACCTAT-1</td>\n",
       "      <td>ACAGCGGGTGCGCGAA-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACAGCCAAACCTTG-1</td>\n",
       "      <td>AAACAGCCAAACCTTG-1</td>\n",
       "      <td>ACAGCGGGTCCTCCAT-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736195</th>\n",
       "      <td>TTTGACTTCCCATAGG-1</td>\n",
       "      <td>TTTGACTTCCCATAGG-1</td>\n",
       "      <td>ACTAGGTAGCTTATCG-1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736196</th>\n",
       "      <td>TTTGACTTCGTGCCTT-1</td>\n",
       "      <td>TTTGACTTCGTGCCTT-1</td>\n",
       "      <td>ACTAGGTAGGACACGC-1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736197</th>\n",
       "      <td>TTTGGTAAGCTTATGA-1</td>\n",
       "      <td>TTTGGTAAGCTTATGA-1</td>\n",
       "      <td>AGATAGCTCACTAATC-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736198</th>\n",
       "      <td>TTTGTGTTCCCATAGG-1</td>\n",
       "      <td>TTTGTGTTCCCATAGG-1</td>\n",
       "      <td>GTTCTCAAGCTTATCG-1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736199</th>\n",
       "      <td>TTTGTGTTCGCCTAAG-1</td>\n",
       "      <td>TTTGTGTTCGCCTAAG-1</td>\n",
       "      <td>GTTCTCAAGTAAAGGG-1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>736200 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   barcode         gex_barcode        atac_barcode  is_cell  \\\n",
       "0       AAACAGCCAAACAACA-1  AAACAGCCAAACAACA-1  ACAGCGGGTGTGTTAC-1        0   \n",
       "1       AAACAGCCAAACATAG-1  AAACAGCCAAACATAG-1  ACAGCGGGTTGTTCTT-1        1   \n",
       "2       AAACAGCCAAACCCTA-1  AAACAGCCAAACCCTA-1  ACAGCGGGTAACAGGC-1        0   \n",
       "3       AAACAGCCAAACCTAT-1  AAACAGCCAAACCTAT-1  ACAGCGGGTGCGCGAA-1        0   \n",
       "4       AAACAGCCAAACCTTG-1  AAACAGCCAAACCTTG-1  ACAGCGGGTCCTCCAT-1        0   \n",
       "...                    ...                 ...                 ...      ...   \n",
       "736195  TTTGACTTCCCATAGG-1  TTTGACTTCCCATAGG-1  ACTAGGTAGCTTATCG-1        0   \n",
       "736196  TTTGACTTCGTGCCTT-1  TTTGACTTCGTGCCTT-1  ACTAGGTAGGACACGC-1        0   \n",
       "736197  TTTGGTAAGCTTATGA-1  TTTGGTAAGCTTATGA-1  AGATAGCTCACTAATC-1        0   \n",
       "736198  TTTGTGTTCCCATAGG-1  TTTGTGTTCCCATAGG-1  GTTCTCAAGCTTATCG-1        0   \n",
       "736199  TTTGTGTTCGCCTAAG-1  TTTGTGTTCGCCTAAG-1  GTTCTCAAGTAAAGGG-1        0   \n",
       "\n",
       "        excluded_reason  gex_raw_reads  gex_mapped_reads  \\\n",
       "0                     0             24                16   \n",
       "1                     0          33621             32008   \n",
       "2                     0             22                19   \n",
       "3                     0             45                41   \n",
       "4                     0             23                22   \n",
       "...                 ...            ...               ...   \n",
       "736195                2              0                 0   \n",
       "736196                2              0                 0   \n",
       "736197                0              0                 0   \n",
       "736198                2              0                 0   \n",
       "736199                2              0                 0   \n",
       "\n",
       "        gex_conf_intergenic_reads  gex_conf_exonic_reads  \\\n",
       "0                               1                     15   \n",
       "1                            3677                   6527   \n",
       "2                               0                     19   \n",
       "3                               0                     35   \n",
       "4                               1                     11   \n",
       "...                           ...                    ...   \n",
       "736195                          0                      0   \n",
       "736196                          0                      0   \n",
       "736197                          0                      0   \n",
       "736198                          0                      0   \n",
       "736199                          0                      0   \n",
       "\n",
       "        gex_conf_intronic_reads  ...  atac_raw_reads  atac_unmapped_reads  \\\n",
       "0                             0  ...               1                    0   \n",
       "1                         19433  ...           16733                  212   \n",
       "2                             0  ...               0                    0   \n",
       "3                             4  ...              27                    1   \n",
       "4                             3  ...               7                    0   \n",
       "...                         ...  ...             ...                  ...   \n",
       "736195                        0  ...              20                    0   \n",
       "736196                        0  ...              23                    0   \n",
       "736197                        0  ...               1                    1   \n",
       "736198                        0  ...              24                    1   \n",
       "736199                        0  ...               4                    1   \n",
       "\n",
       "        atac_lowmapq  atac_dup_reads  atac_chimeric_reads  \\\n",
       "0                  1               0                    0   \n",
       "1               2053            4694                   28   \n",
       "2                  0               0                    0   \n",
       "3                  2               3                    0   \n",
       "4                  1               2                    0   \n",
       "...              ...             ...                  ...   \n",
       "736195             2               5                    0   \n",
       "736196             5               2                    0   \n",
       "736197             0               0                    0   \n",
       "736198             0               6                    0   \n",
       "736199             1               0                    0   \n",
       "\n",
       "        atac_mitochondrial_reads  atac_fragments  atac_TSS_fragments  \\\n",
       "0                              0               0                   0   \n",
       "1                           1836            7901                3107   \n",
       "2                              0               0                   0   \n",
       "3                              0              21                   6   \n",
       "4                              0               4                   0   \n",
       "...                          ...             ...                 ...   \n",
       "736195                         0              13                   0   \n",
       "736196                         0              16                   1   \n",
       "736197                         0               0                   0   \n",
       "736198                         3              14                   0   \n",
       "736199                         0               2                   0   \n",
       "\n",
       "        atac_peak_region_fragments  atac_peak_region_cutsites  \n",
       "0                                0                          0  \n",
       "1                             3586                       6997  \n",
       "2                                0                          0  \n",
       "3                                3                          5  \n",
       "4                                1                          2  \n",
       "...                            ...                        ...  \n",
       "736195                           1                          2  \n",
       "736196                           1                          2  \n",
       "736197                           0                          0  \n",
       "736198                           0                          0  \n",
       "736199                           0                          0  \n",
       "\n",
       "[736200 rows x 31 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_path = \"GSM6284669_S1_per_barcode_metrics.csv.gz\"\n",
    "adata_c2 = pd.read_csv(rna_path, delimiter=',')\n",
    "adata_c2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3422 × 36601\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnadata = sc.read_10x_h5(\"GSM6284671_S1_filtered_feature_bc_matrix.h5\")\n",
    "rnadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>ENSG00000277836</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>ENSG00000276017</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types  genome\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression  GRCh38\n",
       "FAM138A      ENSG00000237613  Gene Expression  GRCh38\n",
       "OR4F5        ENSG00000186092  Gene Expression  GRCh38\n",
       "AL627309.1   ENSG00000238009  Gene Expression  GRCh38\n",
       "AL627309.3   ENSG00000239945  Gene Expression  GRCh38\n",
       "...                      ...              ...     ...\n",
       "AC141272.1   ENSG00000277836  Gene Expression  GRCh38\n",
       "AC023491.2   ENSG00000278633  Gene Expression  GRCh38\n",
       "AC007325.1   ENSG00000276017  Gene Expression  GRCh38\n",
       "AC007325.4   ENSG00000278817  Gene Expression  GRCh38\n",
       "AC007325.2   ENSG00000277196  Gene Expression  GRCh38\n",
       "\n",
       "[36601 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnadata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('halo')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b8353fad07efb87fed686271f8c766cb24580e47ae7007e28a3b0368661d8351"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
